paderwasn/rir_simulation/rir_simulation.py

```diff
@@ -0,0 +1,525 @@
+    and appropriately scaled to match the target DRR.
+
+    Args:
+        rir_ism (array-like):
+            Room impulse response (RIR) consisting of early reflections
+
+        alpha (float):
+            Directivity factor of acoustic source
+
+        distance (float):
+            Distance between microphone position and acoustic source
+
+        directivity_attenuation (float):
+            Source’s directional response based on source directivity,
+            azimuth and elevation of the source
+
+        change_alpha (bool):
+            If true, maximum directivity factor based on early reflections
+            is calculated to guarantee solvability
+
+        lower_limit_alpha (float):
+            Lower limit of the range where alpha factors are drawn from
+
+        All other arguments are described in "calc_rirs".
+
+    Returns:
+        (rir, alpha):
+        rir: None, if no solution could be found for given parameters
+        otherwise combined and scaled RIR
+        alpha: the directivity factor of acoustic source used for generation
+    """
+    # maximum peak of pra is shifted by 40 samples compared to physical
+    # time of arrival (TOA)
+    pra_toa_offset = 40
+
+    # calculate stochastic reverberation part of RIR
+    n = np.arange(rir_len)
+    delta = 3 * np.log(10) / t60
+    toa_direct = int(
+        np.round(distance / sound_speed * sample_rate)) + pra_toa_offset
+    raw_stochastic = np.random.normal(size=rir_len) * np.exp(
+        -delta * (n - toa_direct) / sample_rate)
+    raw_stochastic[:toa_direct + 1] = 0
+
+    lower_lim = toa_direct - win_len_direct
+    upper_lim = toa_direct + win_len_direct + 1
+
+    # adjust alpha to maintain solvability;
+    # in some cases, the RIR of image source method may be already too diffuse
+    # according to target DRR for the currently drawn alpha, so adding
+    # additional reverberant parts causes an invalid solution
+    # if change_alpha is set after some iterations of rejection sampling,
+    # calculate alpha based on the early reflections of the RIR that will
+    # guarantee a solution during scaling to match the target DRR,
+    # but still matches the original range of alpha the best
+    if change_alpha:
+        # calculate DRR of ISM part
+        drr_im = (np.sum(rir_ism[lower_lim:upper_lim] ** 2)
+                  / np.sum(rir_ism[upper_lim:] ** 2))
+        # solve DRR formula to get alpha that would just create the
+        # DRR of the early reflections of the RIR, seen as upper bound for the
+        # maximum DRR to achieve, because only first reflections are simulated
+        # and amending of the stochastic part would overall lower the DRR
+        max_factor = (drr_im * np.pi * t60 * distance ** 2
+                      / np.prod(room_dimensions) * 100
+                      / directivity_attenuation ** 2)
+        if max_factor < alpha:
+            # for current alpha, target DRR would be more direct as DRR of
```

Comment:
For the current value of  alpha, the target DRR would be more smaller than the DRR of